app/src/studio/gatt_rpc_transport.c

```diff
@@ -0,0 +1,222 @@
+    BT_GATT_CHARACTERISTIC(BT_UUID_DECLARE_128(ZMK_STUDIO_BT_RPC_CHRC_UUID),
+                           BT_GATT_CHRC_WRITE | BT_GATT_CHRC_READ | BT_GATT_CHRC_INDICATE,
+                           BT_GATT_PERM_READ_ENCRYPT | BT_GATT_PERM_WRITE_ENCRYPT, read_rpc_resp,
+                           write_rpc_req, NULL),
+    BT_GATT_CCC(rpc_ccc_cfg_changed, BT_GATT_PERM_READ_ENCRYPT | BT_GATT_PERM_WRITE_ENCRYPT));
+
+static void refresh_notify_size(void) {
+    struct bt_conn *conn = zmk_ble_active_profile_conn();
+
+    uint8_t ns = 23; // Default MTU size unless negotiated higher
+    if (conn) {
+
+        struct bt_conn_info conn_info;
+        if (bt_conn_get_info(conn, &conn_info) >= 0) {
+            ns = conn_info.le.data_len->tx_max_len;
+        }
+        bt_conn_unref(conn);
+    }
+
+    atomic_set(&notify_size, ns);
+}
+
+static int gatt_start_rx() {
+    refresh_notify_size();
+    handling_rx = true;
+    return 0;
+}
+
+static int gatt_stop_rx(void) {
+    handling_rx = false;
+    return 0;
+}
+
+static struct bt_gatt_indicate_params rpc_indicate_params = {
+    .attr = &rpc_interface.attrs[1],
+};
+
+static void notif_rpc_tx_cb(struct k_work *work) {
+    struct bt_conn *conn = zmk_ble_active_profile_conn();
+    struct ring_buf *tx_buf = zmk_rpc_get_tx_buf();
+
+    if (!conn) {
+        LOG_WRN("No active connection for queued data, dropping");
+        ring_buf_reset(tx_buf);
+        return;
+    }
+
+    uint8_t notify_size = 23; // Default MTU size unless negotiated higher
+    struct bt_conn_info conn_info;
+    if (bt_conn_get_info(conn, &conn_info) >= 0) {
+        notify_size = conn_info.le.data_len->tx_max_len;
+    }
+
+    uint8_t notify_bytes[notify_size];
+
+    while (ring_buf_size_get(tx_buf) > 0) {
+        uint8_t added = 0;
+        while (added < notify_size && ring_buf_size_get(tx_buf) > 0) {
+            uint8_t *buf;
+            int len = ring_buf_get_claim(tx_buf, &buf, notify_size - added);
+
+            if (len > 0) {
+                memcpy(notify_bytes + added, buf, len);
+            }
+
+            added += len;
+            ring_buf_get_finish(tx_buf, len);
+        }
+
+        rpc_indicate_params.data = notify_bytes;
+        rpc_indicate_params.len = added;
+
+        uint8_t notify_attempts = 5;
+        do {
+            int err = bt_gatt_indicate(conn, &rpc_indicate_params);
+            if (err >= 0) {
+                break;
+            }
+
+            LOG_WRN("Failed to notify the response %d", err);
+            k_sleep(K_MSEC(200));
+        } while (notify_attempts-- > 0);
+    }
+
+    bt_conn_unref(conn);
+}
+
+static K_WORK_DEFINE(notify_tx_work, notif_rpc_tx_cb);
+
+struct gatt_write_state {
+    uint8_t pending_notify;
+};
+
+static void gatt_tx_notify(struct ring_buf *tx_buf, size_t added, bool msg_done, void *user_data) {
+    struct gatt_write_state *state = (struct gatt_write_state *)user_data;
+
+    state->pending_notify += added;
```

Comment:
We are writing a size_t to a uint8_t. Can we guarantee that this is always <= 255?

Reply:
Not 100%. Updated to just store a `size_t` in the state.